server/src/main/java/org/elasticsearch/search/aggregations/pipeline/MovFnPipelineAggregator.java

```diff
@@ -117,30 +119,53 @@ public InternalAggregation reduce(InternalAggregation aggregation, InternalAggre
 
         MovingFunctionScript executableScript = scriptFactory.newInstance();
 
+        List<Double> values = buckets.stream()
+            .map(b -> resolveBucketValue(histo, b, bucketsPaths()[0], gapPolicy))
+            .filter(v -> v != null && !v.isNaN())
+            .collect(Collectors.toList());
+
+        int index = 0;
         for (InternalMultiBucketAggregation.InternalBucket bucket : buckets) {
             Double thisBucketValue = resolveBucketValue(histo, bucket, bucketsPaths()[0], gapPolicy);
 
             // Default is to reuse existing bucket.  Simplifies the rest of the logic,
             // since we only change newBucket if we can add to it
             MultiBucketsAggregation.Bucket newBucket = bucket;
 
-            if (thisBucketValue != null && thisBucketValue.equals(Double.NaN) == false) {
+            if (thisBucketValue != null && !thisBucketValue.isNaN()) {
```

Comment:
Just an FYI style nit: we avoid using the short `!foo` negation since it's easy to miss, and instead use the longer form `foo == false` or `foo != true`.